### PR TITLE
Add DoFPartitioner

### LINKED ISSUE #13104: Partitioner with weights according to DOFs on each block
## Reason

Better load balancing for problems with varying DOFs per block.

## Design

Inherit from `PetscPartitioner` and override weight assignment.

## Impact

New capability

### PR DESCRIPTION
**Pull request by snschune**
closes #13104

### CODE REVIEW

**Comment by snschune** [comment]
Something is screwed up and I can't figure out what. When I try to run the test, it tells me `Unable to suppress nonexistent parameter: apply_element_weight` and I have no idea why it would do this. Somehow `DoFPartitioner` does not get the parameters from `PetscExternalPartitioner`. If I comment the lines in `validParams` out, it breaks on a bad cast when `clone()` is called. In that case, it appears I somehow do not inherit from `PetscExternalPartitioner`.

What am I missing here @fdkong @friedmud ?

**Comment by YaqiWang** [comment]
I do not understand the error. But one thing that could be problematic is, `elem.n_dofs` may not be available before mesh is partitioned.

**Comment by snschune** [comment]
That's what I get (see tests):
```
partitioners/petsc_partitioner.dof: Working Directory: /opt/civet/build_0/moose/test/tests/partitioners/petsc_partitioner
partitioners/petsc_partitioner.dof: Running command: mpiexec -n 4 /opt/civet/build_0/moose/test/moose_test-dbg -i dof_partitioner.i --error --error-unused --error-override --no-gdb-backtrace
partitioners/petsc_partitioner.dof: 
partitioners/petsc_partitioner.dof: 
partitioners/petsc_partitioner.dof: 
partitioners/petsc_partitioner.dof: 
partitioners/petsc_partitioner.dof: 
partitioners/petsc_partitioner.dof: *** ERROR ***
partitioners/petsc_partitioner.dof: Unable to suppress nonexistent parameter: apply_element_weight
partitioners/petsc_partitioner.dof: 
partitioners/petsc_partitioner.dof: 
partitioners/petsc_partitioner.dof: application called MPI_Abort(MPI_COMM_WORLD, 1) - process 2
partitioners/petsc_partitioner.dof: 
partitioners/petsc_partitioner.dof: 
partitioners/petsc_partitioner.dof: 
partitioners/petsc_partitioner.dof: 
partitioners/petsc_partitioner.dof: *** ERROR ***
partitioners/petsc_partitioner.dof: Unable to suppress nonexistent parameter: apply_element_weight
partitioners/petsc_partitioner.dof: 
partitioners/petsc_partitioner.dof: 
partitioners/petsc_partitioner.dof: application called MPI_Abort(MPI_COMM_WORLD, 1) - process 1
partitioners/petsc_partitioner.dof: 
partitioners/petsc_partitioner.dof: ################################################################################
partitioners/petsc_partitioner.dof: Tester failed, reason: ERRMSG
partitioners/petsc_partitioner.dof: 
```

File: framework/src/partitioner/DoFPartitioner.C [hunk 1]
```diff
@@ -0,0 +1,44 @@
...
+
+registerMooseObject("MooseApp", DoFPartitioner);
+
+template <>
+InputParameters
+validParams<DoFPartitioner>()
+{
+  InputParameters params = validParams<PetscExternalPartitioner>();
+  params.set<bool>("apply_element_weight") = true;
+  params.suppressParameter<Real>("apply_element_weight");
```
Comment (fdkong) [inline framework/src/partitioner/DoFPartitioner.C]: Real should not be "bool" here?

**Review by fdkong — CHANGES_REQUESTED** [review]

**Comment by snschune** [comment]
@fdkong thanks! that was a pretty dumb one.  But here is the error I was talking about earlier:

```
partitioners/petsc_partitioner.dof_parallel_mesh: Working Directory: /opt/civet/build_0/moose/test/tests/partitioners/petsc_partitioner
partitioners/petsc_partitioner.dof_parallel_mesh: Running command: mpiexec -n 4 /opt/civet/build_0/moose/test/moose_test-opt -i dof_partitioner.i Mesh/parallel_type=distributed Outputs/file_base=dof_partitioner_parallel_mesh_out --error --error-unused --error-override --no-gdb-backtrace
partitioners/petsc_partitioner.dof_parallel_mesh: 
partitioners/petsc_partitioner.dof_parallel_mesh: 
partitioners/petsc_partitioner.dof_parallel_mesh: 
partitioners/petsc_partitioner.dof_parallel_mesh: 
partitioners/petsc_partitioner.dof_parallel_mesh: 
partitioners/petsc_partitioner.dof_parallel_mesh: *** ERROR ***
partitioners/petsc_partitioner.dof_parallel_mesh: std::bad_cast
partitioners/petsc_partitioner.dof_parallel_mesh: 
partitioners/petsc_partitioner.dof_parallel_mesh: 
partitioners/petsc_partitioner.dof_parallel_mesh: 
partitioners/petsc_partitioner.dof_parallel_mesh: 
partitioners/petsc_partitioner.dof_parallel_mesh: 
partitioners/petsc_partitioner.dof_parallel_mesh: 
partitioners/petsc_partitioner.dof_parallel_mesh: *** ERROR ***
partitioners/petsc_partitioner.dof_parallel_mesh: std::bad_cast
partitioners/petsc_partitioner.dof_parallel_mesh: 
partitioners/petsc_partitioner.dof_parallel_mesh: 
partitioners/petsc_partitioner.dof_parallel_mesh: 
partitioners/petsc_partitioner.dof_parallel_mesh: 
partitioners/petsc_partitioner.dof_parallel_mesh: 
partitioners/petsc_partitioner.dof_parallel_mesh: 
partitioners/petsc_partitioner.dof_parallel_mesh: *** ERROR ***
partitioners/petsc_partitioner.dof_parallel_mesh: std::bad_cast
partitioners/petsc_partitioner.dof_parallel_mesh: 
partitioners/petsc_partitioner.dof_parallel_mesh: 
partitioners/petsc_partitioner.dof_parallel_mesh: 
partitioners/petsc_partitioner.dof_parallel_mesh: 
partitioners/petsc_partitioner.dof_parallel_mesh: 
partitioners/petsc_partitioner.dof_parallel_mesh: 
partitioners/petsc_partitioner.dof_parallel_mesh: *** ERROR ***
partitioners/petsc_partitioner.dof_parallel_mesh: std::bad_cast
partitioners/petsc_partitioner.dof_parallel_mesh: 
partitioners/petsc_partitioner.dof_parallel_mesh: 
partitioners/petsc_partitioner.dof_parallel_mesh: application called MPI_Abort(MPI_COMM_WORLD, 1) - process 3
partitioners/petsc_partitioner.dof_parallel_mesh: application called MPI_Abort(MPI_COMM_WORLD, 1) - process 1
partitioners/petsc_partitioner.dof_parallel_mesh: application called MPI_Abort(MPI_COMM_WORLD, 1) - process 0
partitioners/petsc_partitioner.dof_parallel_mesh: application called MPI_Abort(MPI_COMM_WORLD, 1) - process 2
partitioners/petsc_partitioner.dof_parallel_mesh: 
partitioners/petsc_partitioner.dof_parallel_mesh: ################################################################################
```

It only happens with `parallel_type=distributed`.

File: framework/src/partitioner/DoFPartitioner.C [hunk 2]
```diff
@@ -0,0 +1,44 @@
...
+std::unique_ptr<Partitioner>
+DoFPartitioner::clone() const
+{
+  return libmesh_make_unique<DoFPartitioner>(_pars);
+}
+
+dof_id_type
+DoFPartitioner::computeElementWeight(Elem & elm)
+{
+  return elm.n_dofs(_app.getExecutioner()->feProblem().getNonlinearSystemBase().number());
```
Comment (snschune) [inline framework/src/partitioner/DoFPartitioner.C]: @fdkong before merging this, I should make sure that `n_dofs` is available at partitioning.

**Comment by fdkong** [comment]
> @fdkong thanks! that was a pretty dumb one. But here is the error I was talking about earlier:
> 
> ```
> partitioners/petsc_partitioner.dof_parallel_mesh: Working Directory: /opt/civet/build_0/moose/test/tests/partitioners/petsc_partitioner
> partitioners/petsc_partitioner.dof_parallel_mesh: Running command: mpiexec -n 4 /opt/civet/build_0/moose/test/moose_test-opt -i dof_partitioner.i Mesh/parallel_type=distributed Outputs/file_base=dof_partitioner_parallel_mesh_out --error --error-unused --error-override --no-gdb-backtrace
> partitioners/petsc_partitioner.dof_parallel_mesh: 
> partitioners/petsc_partitioner.dof_parallel_mesh: 
> partitioners/petsc_partitioner.dof_parallel_mesh: 
> partitioners/petsc_partitioner.dof_parallel_mesh: 
> partitioners/petsc_partitioner.dof_parallel_mesh: 
> partitioners/petsc_partitioner.dof_parallel_mesh: *** ERROR ***
> partitioners/petsc_partitioner.dof_parallel_mesh: std::bad_cast
> partitioners/petsc_partitioner.dof_parallel_mesh: 
> partitioners/petsc_partitioner.dof_parallel_mesh: 
> partitioners/petsc_partitioner.dof_parallel_mesh: 
> partitioners/petsc_partitioner.dof_parallel_mesh: 
> partitioners/petsc_partitioner.dof_parallel_mesh: 
> partitioners/petsc_partitioner.dof_parallel_mesh: 
> partitioners/petsc_partitioner.dof_parallel_mesh: *** ERROR ***
> partitioners/petsc_partitioner.dof_parallel_mesh: std::bad_cast
> partitioners/petsc_partitioner.dof_parallel_mesh: 
> partitioners/petsc_partitioner.dof_parallel_mesh: 
> partitioners/petsc_partitioner.dof_parallel_mesh: 
> partitioners/petsc_partitioner.dof_parallel_mesh: 
> partitioners/petsc_partitioner.dof_parallel_mesh: 
> partitioners/petsc_partitioner.dof_parallel_mesh: 
> partitioners/petsc_partitioner.dof_parallel_mesh: *** ERROR ***
> partitioners/petsc_partitioner.dof_parallel_mesh: std::bad_cast
> partitioners/petsc_partitioner.dof_parallel_mesh: 
> partitioners/petsc_partitioner.dof_parallel_mesh: 
> partitioners/petsc_partitioner.dof_parallel_mesh: 
> partitioners/petsc_partitioner.dof_parallel_mesh: 
> partitioners/petsc_partitioner.dof_parallel_mesh: 
> partitioners/petsc_partitioner.dof_parallel_mesh: 
> partitioners/petsc_partitioner.dof_parallel_mesh: *** ERROR ***
> partitioners/petsc_partitioner.dof_parallel_mesh: std::bad_cast
> partitioners/petsc_partitioner.dof_parallel_mesh: 
> partitioners/petsc_partitioner.dof_parallel_mesh: 
> partitioners/petsc_partitioner.dof_parallel_mesh: application called MPI_Abort(MPI_COMM_WORLD, 1) - process 3
> partitioners/petsc_partitioner.dof_parallel_mesh: application called MPI_Abort(MPI_COMM_WORLD, 1) - process 1
> partitioners/petsc_partitioner.dof_parallel_mesh: application called MPI_Abort(MPI_COMM_WORLD, 1) - process 0
> partitioners/petsc_partitioner.dof_parallel_mesh: application called MPI_Abort(MPI_COMM_WORLD, 1) - process 2
> partitioners/petsc_partitioner.dof_parallel_mesh: 
> partitioners/petsc_partitioner.dof_parallel_mesh: ################################################################################
> ```
> It only happens with `parallel_type=distributed`.

It is a bug in GeneratedMeshGenerator.  Fixed at 6dae5ef. It is not related to the partitioner

You may also take @YaqiWang's comments into consideration. The DoFs are not ready yet when doing partitioning. You may need some kind of two-stages partitions. Partition the mesh without DoFs information, build DoFs, and then partition the mesh again with DoFs information. It is just a egg-chicken problem.

File: framework/src/partitioner/DoFPartitioner.C (hunk 2 above)
Comment (fdkong) [reply framework/src/partitioner/DoFPartitioner.C]: See my comments below.

**Comment by snschune** [comment]
This PR will require to split mesh reading and partitioning into separate tasks. The variable block info is added after the mesh is read, but before partitioning, see

https://github.com/idaholab/moose/blob/4cba731d990eb6692ecd6b81d9686f9f65053758/framework/src/base/Moose.C#L242

@fdkong please ping me once that is done. 

I assume that certain data structures in the `MeshBase` are initialized before partitioning, e.g. 

https://github.com/idaholab/moose/blob/4cba731d990eb6692ecd6b81d9686f9f65053758/framework/src/actions/AddBlockVariableInfoAction.C#L97

which relies on `MooseMesh::getSubdomainID(const SubdomainName & subdomain_name) const`
and that ultimately requires `MeshBase::_block_id_to_name` to be set up. That should happen either when exodus files are read or - by changes in this PR - at task `set_mesh_base`.

I also require access to element iterators here 
https://github.com/idaholab/moose/blob/4cba731d990eb6692ecd6b81d9686f9f65053758/framework/src/actions/AddBlockVariableInfoAction.C#L74
I assume that would be set up before partitioning @fdkong ?

**Pull request closed without merging**